### PR TITLE
win32/meson: build as console program, use GUI wrapper

### PR DESCRIPTION
**Pull request by bitraid**
>Having the main mpv executable be a console program (mpv.com) and using a GUI wrapper (mpv.exe) to start without a console, is simpler than having a GUI program with a console wrapper. There is no need for using environment variables, attaching to parent console, etc.
>
>All mpv has to do, is to enable the Peudo-GUI profile when needed. This is now done when not attached to a console, but also when it’s the only attached process in the console (e.g., started from Explorer, the Run dialog, or using the `start` command. In this case it also passes `--terminal`).
>
>The wrapper starts mpv detached form console, and copies the window show state and any inherited stdio handles.

This might be a controversial change, and I understand that it might not get accepted. Nevertheless, I think that having the main executable be a console program, fits better mpv as a “command line video player”. But the main benefit, is that it simplifies the code for the CLI/GUI subsystem workaround.

This change barely introduces new code, while it eliminates some hacks. The GUI wrapper is basically the console wrapper with some minor changes. The only drawback that I can think of, is that the main program would now have the `.com` extension.

There is only one commit because everything depends on each other.

### CODE REVIEW

**Comment by Dudemanguy** [comment]
No strong opinions on the particular details of the windows implementation (I'll leave that to others), but as long as there's still an exe for windows users to click on with the pseudo-gui mode, it's probably OK.

**Comment by kasper93** [comment]
It has been like that for years, that's why I didn't touch it yet. In my opinion wrapper should exist, if we start .exe from console it should attach to existing one, else just use gui mode.

Back to this PR, I don't like an idea of having main executable as `mpv.com`.

**Comment by bitraid** [comment]
> In my opinion wrapper shouldn't exist, if we start .exe from console it should attach to existing one, else just use gui mode.

But when a GUI program is started from the console, the process is forked and the control returns immediately to the caller. That means that both programs would use the console at the same time (there are programs that do that, and it's a mess). And that's the main purpose of the console wrapper: to wait for the GUI program to end. It could've been designed differently though, so that mpv didn't need to attach to the console: start the GUI program with anonymous pipes as stdout/stderr, and write to the console whatever comes through. I've written something like this a few years back (just a POC in 32-bit assembly - complies with fasm: [conwrap.asm](https://github.com/bitraid/conwrap/blob/master/conwrap.asm) ).

> 
> Back to this PR, I don't like an idea of having main executable as `mpv.com`.
> 
> Also current way follows established trend of using thing .com wrapper for .exe application. Like Microsoft does for devenv.com and devenv.exe.

`devenv` is basically GUI though. For mpv makes sense to behave as console program by default. But I get why the `.com` extension might not be liked. Personally, I would prefer if there were a console `mpv.exe`, and a GUI `mpvw.exe` (for shortcuts and file associations), but that would mess with existing setups.

File: osdep/main-fn-win.c [hunk 1]
```diff
@@ -80,8 +70,10 @@ int main(int argc_, char **argv_)
     if (argc > 0 && argv[0])
         MP_TARRAY_APPEND(NULL, argv_u8, argv_len, mp_to_utf8(argv_u8, argv[0]));
     if (gui) {
-        MP_TARRAY_APPEND(NULL, argv_u8, argv_len,
-                         "--player-operation-mode=pseudo-gui");
+        MP_TARRAY_APPEND(NULL, argv_u8, argv_len, "--player-operation-mode=pseudo-gui");
+        // Enable terminal output if attached to a console
+        if (cproc_count)
+            MP_TARRAY_APPEND(NULL, argv_u8, argv_len, "--terminal=yes");
```
Comment (kasper93) [inline osdep/main-fn-win.c:76]: Wouldn't it break `--terminal=no` from user command line?

**Comment by kasper93** [comment]
> But when a GUI program is started from the console, the process is forked and the control returns immediately to the caller.

Was more thinking about console mode by default and hide console if not needed, but this obviously introduces a flicker console that is opened/closed.

In general this is not supported case by Windows. I think our current solution is reasonable elegant.

One thing that could be saner is to not interact between mpv.com and mpv.exe. Link them both to shared lib and have them as thin launchers of console and gui mode respectively. Of course this doesn't work for static linking everything in one binary, because it would duplicate the code in both. I don't like the obsession with static linking everything tho.

>  I would prefer if there were a console mpv.exe, and a GUI mpvw.exe (for shortcuts and file associations), but that would mess with existing setups.

`.com` precedence over `.exe` make it quite convenient for both to coexists. 

> But the main benefit, is that it simplifies the code for the CLI/GUI subsystem workaround.

To be honest I don't see it. There is still much code to handle it just different than previous. We can argue which approach is better, but in reality we still doing the workaround to handle both CLI/GUI mode.

File: osdep/main-fn-win.c (hunk 1 above)
Comment (bitraid) [reply osdep/main-fn-win.c:76]: No. User options are added last, so they're the ones that will ultimately be applied.

**Comment by llyyr** [comment]
Does Freesync/G-Sync on Windows work if the mpv process is a .com file and not an .exe file?

**Pull request closed without merging**